File: _CodeTopics/LeetCode_contest/biweekly/biweekly2021/67/67_2_ugly.py
class Solution(object):
    def goodDaysToRobBank(self, security, time):
        """
        :type security: List[int]
        :type time: int
        :rtype: List[int]
        """
        
        n = len(security)
        if n < 2 * time + 1:
            return []
        
        leftinds = []
        leftind1 = 0
        inWindowState = False
        while leftind1 < n - 2*time:
            leftind2 = leftind1 + time
            if not inWindowState:
                breakflag = False
                for i in range(leftind1, leftind2):
                    if security[i] >= security[i+1]:
                        continue
                    else:
                        breakflag = True
                        leftind1 = i + 1
                        break
                if not breakflag:
                    inWindowState = True
                    leftinds.append(leftind1)
                    leftind1 += 1
            else:
                if security[leftind2 - 1] >= security[leftind2]:
                    leftinds.append(leftind1)
                    leftind1 += 1
                else:
                    inWindowState = False
                    leftind1 = leftind2
        ##print leftinds

        rightinds = []
        rightind1 = time
        inWindowState = False
        while rightind1 < n - time:
            rightind2 = rightind1 + time
            if not inWindowState:
                breakflag = False
                for i in range(rightind1, rightind2):
                    if security[i] <= security[i+1]:
                        continue
                    else:
                        breakflag = True
                        rightind1 = i + 1
                        break
                if not breakflag:
                    inWindowState = True
                    rightinds.append(rightind1 + 1)
                    rightind1 += 1
            else:
                if security[rightind2 - 1] <= security[rightind2]:
                    rightinds.append(rightind1 + 1)
                    rightind1 += 1
                else:
                    inWindowState = False
                    rightind1 = rightind2
        ##print rightinds        
        
        rightinds = set(rightinds)
        res = []
        for ind in leftinds:
            if ind + time + 1 in rightinds:
                res.append(ind + time)
        return res
    
"""
https://leetcode-cn.com/submissions/detail/247596451/

131 / 131 个通过测试用例
状态：通过
执行用时: 5416 ms
内存消耗: 31.3 MB
"""

File: _CodeTopics/LeetCode/2001-2200/002100--biweekly67_2/tran/67_2_ugly.cpp
class Solution {
public:
    vector<int> goodDaysToRobBank(vector<int>& security, int time) {

        // 第 67 场双周赛第二题
        // https://leetcode-cn.com/submissions/detail/247596451/ --> 对着这个翻译完，自己都晕了。

        int n = security.size();
        if (n < 2 * time + 1)
            return vector<int>{};
        
        vector<int> leftinds = {};
        int leftind1 = 0;
        bool inWindowState = false;
        while (leftind1 < n - 2*time) {
            int leftind2 = leftind1 + time;
            if (!inWindowState) {
                bool breakflag = false;
                for (int i = leftind1; i < leftind2; ++i) {
                    if (security[i] >= security[i+1]) {
                        continue;
                    } else {
                        breakflag = true;
                        leftind1 = i + 1;
                        break;
                    }
                }
                if (!breakflag) {
                    inWindowState = true;
                    leftinds.push_back(leftind1);
                    ++leftind1;
                }
            } else {
                if (security[leftind2 - 1] >= security[leftind2]) {
                    leftinds.push_back(leftind1);
                    ++leftind1;
                } else {
                    inWindowState = false;
                    leftind1 = leftind2;
                }
            }
        }

        vector<int> rightinds = {};
        int rightind1 = time;
        inWindowState = false;
        while (rightind1 < n - time) {
            int rightind2 = rightind1 + time;
            if (!inWindowState) {
                bool breakflag = false;
                for (int i = rightind1; i < rightind2; ++i) {
                    if (security[i] <= security[i+1]) {
                        continue;
                    } else {
                        breakflag = true;
                        rightind1 = i + 1;
                        break;
                    }
                }
                if (!breakflag) {
                    inWindowState = true;
                    rightinds.push_back(rightind1 + 1);
                    ++rightind1;
                }
            } else {
                if (security[rightind2 - 1] <= security[rightind2]) {
                    rightinds.push_back(rightind1 + 1);
                    ++rightind1;
                } else {
                    inWindowState = false;
                    rightind1 = rightind2;
                }
            }
        }

        unordered_set<int> tmp_rightinds = unordered_set<int>(rightinds.begin(), rightinds.end());
        vector<int> res;
        for (int ind : leftinds) {
            if (tmp_rightinds.find(ind + time + 1) != tmp_rightinds.end()) {
                res.push_back(ind + time);
            }
        }
        return res;
    }
};

/*
https://leetcode-cn.com/submissions/detail/277985399/

执行用时：144 ms, 在所有 C++ 提交中击败了12.60%的用户
内存消耗：98.3 MB, 在所有 C++ 提交中击败了5.08%的用户
通过测试用例：
144 / 144
*/
